### DISCUSSION TITLE
Updating Kernel and applying updates

### DISCUSSION THREAD

**Discussion by Shinmegumi** [jglathe/linux_ms_dev_kit#10] [General]
I just came across this for the dev kit. I've had one for quite some time and wanted to get Linux running on it so I could tinker with emulation and Wine and such. I was wondering how easy it was to apply a new kernel over this now that there is more mainline support for the WDK 2023. Do I still need to follow the instructions on the main page? I'm not wholly familiar with Linux and have never really compiled a kernel before.

**Comment by mirostauder** [comment]
I believe the instructions are outdated.

You can use latest rolling release of Armbian
https://github.com/armbian/os/releases - search for Wdk2023 images.
These are rebuild twice a day, so get the latest.
So far Armbian for Wdk2023 uses **6.7.0-rc6** kernel - which works for me.

If you want the latest kernel from @jglathe, use the [jg/blackrock-v6.7.y](https://github.com/jglathe/linux_ms_dev_kit/tree/jg/blackrock-v6.7.y) branch
currently **6.7.3-g5245a47e48fa** - you need to compile yourself.

I wrote this script to automate the build/install:
```
#!/usr/bin/env bash

#BRANCH="jg/blackrock-v6.8.y"
#BRANCH="jg/mainline"
BRANCH="jg/v6.9-rc5"

umount linux_ms_dev_kit &>/dev/null
rm -rf linux_ms_dev_kit &>/dev/null
mkdir linux_ms_dev_kit &>/dev/null
mount -t tmpfs tmpfs linux_ms_dev_kit

git clone --depth 1 -b ${BRANCH} https://github.com/jglathe/linux_ms_dev_kit.git linux_ms_dev_kit
cd linux_ms_dev_kit

make -j$(nproc) devkit_defconfig
sed -i 's/.*CONFIG_LOCALVERSION_AUTO.*/CONFIG_LOCALVERSION_AUTO=y/g' .config

VERS=$(make kernelversion)
HASH=$(git rev-parse HEAD)

make -j$(nproc)

make -j$(nproc) modules_install

mkdir -p /usr/lib/linux-image-${VERS}-g${HASH}/qcom/
cp ./arch/arm64/boot/dts/qcom/sc8280xp-microsoft-blackrock.dtb /usr/lib/linux-image-${VERS}-g${HASH}/qcom/sc8280xp-microsoft-dev-kit-2023.dtb
image-${VERS}${CONFLOCVER}${LOCALVERSION}/qcom/sc8280xp-microsoft-dev-kit-2023.dtb
make -j$(nproc) install

rm /boot/*.old &>/dev/null && update-grub && sync
echo "Done in $SECONDS sec."
```
*FIX - added missing `update-grub`, thanks @jglathe 
*2024-04-27 UPDATE - BRANCH variable, for easy changing
*2024-04-27 UPDATE - use TMPFS, for build speedup

To build the kernel, you will need to install `build-essential` and some more stuff...
just install missing stuff as you get build errors - rinse/repeat.

EDIT:
from the front page, build dependencies:
`apt install git bc bison flex libssl-dev make libc6-dev libncurses5-dev build-essential`

**Reply by jglathe** [reply]
Nice script. :grin: After deleting the .old entries you need to do another update-grub IMO. Otherwise really nice for armbian.
Newest kernel is 6.8.0-rc5. Will also have a look at 6.7.5. 
I still try to get a 6.7 image out - experimented a bit with the official Ubuntu x13s install image, but that one's tedious, too. 
Main changes will be:
- updated Ubuntu 23.10
- newer kernel (whatever looks rock-solid)
- use [bootmac](https://gitlab.com/postmarketOS/bootmac/-/merge_requests/7)  for setting up "unique" and stable mac addresses on WLAN and Bluetooth
- most recent WLAN firmware

**Reply by Shinmegumi** [reply]
I'll keep an eye out for the 6.7 image. I'm just looking to play around and tinker with my dev kit in a way to utilize it for emulation and just ARM performance testing with Linux. I've been curious to see what Snapdragon's SoC is capable of and Windows on ARM didn't use an updated driver for the GPU to get the most out of it.

**Comment by Shinmegumi** [comment]
Thank you for the information. I don't know the difference between armbian and something like Ubuntu 23.10, but I was looking at installing a full desktop distro like Ubuntu. I'm guessing these steps would not be too different?

**Reply by jglathe** [reply]
The build instructions should still work. The [Ubuntu 23.10 image](https://drive.google.com/file/d/1bVxz3DLN8otFIQsVYyWqC7DhJhawdI1i/view?usp=sharing) is still a good choice, you get full desktop OS. Performance ist quite impressive IMO. Regarding Wine... no idea how it is on an aarch64 machine, and then Qualcomm, too (no EL2, no /dev/kvm). Usually you're limited to containers, no VMs. So not the same breadth of use cases available, but still a lot you can do (with little power budget). 
I am testing EL2 on one box, however... 😈 And yes I have /dev/kvm there, and since this weekend also PCIe. But no display atm.

**Reply by mirostauder** [reply]
@jglathe
yes, yes, yes for KVM!

**Reply by jglathe** [reply]
Interesting timing here. EL2 on newer QualComm chips is made possible through [slbounce](https://github.com/TravMurav/slbounce), quite an ingenious solution IMO. I have a branch for EL2 building on this here, but its hackery. Nevertheless:
```
jglathe@sdbox2:/mnt/Work$ uname -a
Linux sdbox2 6.8.0-rc5-EL2-IOMMU+ #11 SMP PREEMPT Mon Feb 19 20:57:01 CET 2024 aarch64 aarch64 aarch64 GNU/Linux
jglathe@sdbox2:/mnt/Work$ lxc list
+--------+---------+-----------------------+-------------------------------------------------+-----------------+-----------+
|  NAME  |  STATE  |         IPV4          |                      IPV6                       |      TYPE       | SNAPSHOTS |
+--------+---------+-----------------------+-------------------------------------------------+-----------------+-----------+
| ubuntu | RUNNING | 10.225.42.42 (enp5s0) | fd42:d39b:8c86:b2f1:216:3eff:feb8:6081 (enp5s0) | VIRTUAL-MACHINE | 0         |
+--------+---------+-----------------------+-------------------------------------------------+-----------------+-----------+
```
To get PCIe usable on EL2, Linux needs to control the arm-smmuv3. Which works after you get the mappings in the dts right. But collides with the arm-smmuv2 which does the non-pcie stuff on the SoC. Found out at the weekend that this was the problem, and... 6.8 is working on enabling multiple IOMMU drivers. This already works to the point where you need graphics memory, but I'm quite sure this will progress pretty fast. Then there is the issue of loading firmwares for coprocessors without HYP support (which gets killed by slbounce), but one can do nicely on Volterra without it until its figured out. So you're getting 8 cores and 32gigs of RAM for VM/container stuff. And working thermal management.

**Comment by mirostauder** [comment]
@Shinmegumi 
I don't think plain Ubuntu ARM image is bootable on WDK2023,
you would need to splice in a custom kernel/initrd/dtb

Armbian focuses on non-PC devices (SBCs, Tablets, Phones, etc..),
optimizes the distro specifically for the respective device.
https://www.armbian.com/

There are 2 versions of Armbian, Ubuntu based and Debian based.
Mostly the same stuff, depends on your preferences.
for Ubuntu base choose release - **Noble**
for Debian base choose release - **Trixie**

**Reply by jglathe** [reply]
WDK2023 is not in mainline, but a patch for it is prepared and awaiting action/approval from the original author @merckhung.

**Comment by jglathe** [comment]
#11

**Comment by jglathe** [comment]
There is now a [how-to](https://github.com/jglathe/linux_ms_dev_kit/wiki/Booting-up-Qualcomm-SC8280XP-with-Linux-on-EL2) to bring a box up with working KVM support. Image with 6.8 will follow soon-ish.

**Comment by jglathe** [comment]
Kernel branch [v6.8.y](https://github.com/jglathe/linux_ms_dev_kit/tree/jg/blackrock-v6.8.y) is now on v6.8.1. Same with the dedicated [EL2](https://github.com/jglathe/linux_ms_dev_kit/tree/jg/el2-blackrock-v6.8.y) branch. I guess I will do an updated image with this kernel, seems to work smoothly. 
Also, I'm slowly coming along with this debian stuff. Building installable deb packages for the kernel shouldn't be too far away, and also debootstrap doesn't look too strange now.
